src/klein/app.py

```diff
@@ -186,9 +194,37 @@ def index(request):
 
         @returns: decorated handler function.
         """
-        segment_count = url.count('/')
-        if url.endswith('/'):
-            segment_count -= 1
+
+        def _urlToString(_url):
+            return _url.asText() if URL is not False else _url
+
+        if URL is not False:
+
+            if isinstance(url, string_types + (binary_type, )):
+                url = URL.fromText(url)
+            if url.absolute or url.port or url.userinfo:
+                warnings.warn(
+                    "Mind to use base URI for an application routing, "
+                    "instead of "
+                    + url.asText(includeSecrets=False))
+                url = url.replace(
+                    host=None,
+                    scheme=None,
+                    port=None,
+                    userinfo=u"",
+                )
+
+            if not url.rooted:
+                url = url.replace(rooted=True)
```

Comment:
So… none of this checking is happening in the case where we don't have `URL`, which means that we have different behavior for even routes specified with text depending on whether the URL module is present or not; that seems possibly confusing.

There's no attempt to prefix url with "/" in the else clause below, for example, and no complaints if the URL includes non-path components.

Do we need to be doing any of that here, or does `werkzeug` do that?